### ISSUE TITLE
Segmentation fault on reading named columns with a spatial mask

### ISSUE THREAD

**Issue by kyleaoman** [SWIFTSIM/swiftsimio#127]
There's an upstream issue (https://github.com/h5py/h5py/issues/1773) in `h5py` that seems to affect `python3.6` with `h5py==3.0.0` and `h5py==3.1.0`. Earlier `h5py` verisons (`2.x`) are not affected, and the issue is apparently resolved in forthcoming version (https://github.com/h5py/h5py/pull/1796). It might also affect other python versions, but so far `python3.7`, `3.8`, `3.9` and `3.10` haven't crashed with segfaults for me...

The upstream bug is easy to trigger:

```python
import h5py
import numpy as np

h5_filename = "bug.hdf5"
N = 16

with h5py.File(h5_filename, "w") as f:
    data = np.random.rand(2 * N).reshape((N, 2))
    f.create_dataset(
        "dset",
        data=data,
        dtype=float
    )

with h5py.File(h5_filename, "r") as f:
    handle = f["/dset"]
    output = np.empty((N, ), dtype=float)
    source_sel = np.s_[:, 0]
    dest_sel = np.s_[:]
    assert output.flags["C_CONTIGUOUS"]
    assert handle[source_sel].shape == output[dest_sel].shape
    # segmentation fault:
    handle.read_direct(output, source_sel=source_sel, dest_sel=dest_sel)
```

Here's a script that makes this crop up in swiftsimio when reading a namedcolumn dataset:

```python
import os
import h5py
import numpy as np
import unyt as u
from swiftsimio import Writer, SWIFTDataset, SWIFTMask
from swiftsimio.units import cosmo_units

toysnap_filename = "ssio_bug.hdf5"
present_particle_types = {0: "gas", }
boxsize = 10.0
n_g = 32**3


def create_toysnap(snapfile):
    """
    Creates a simple snapshot file.
    """

    # Fill in generic gas quantities.
    sd = Writer(cosmo_units, np.ones(3, dtype=float) * boxsize * u.Mpc)
    getattr(sd, present_particle_types[0]).coordinates = \
        np.random.rand(n_g, 3) * 10 * u.Mpc
    getattr(sd, present_particle_types[0]).velocities = \
        (np.random.rand(n_g, 3) * 2 - 1) * u.km / u.s
    getattr(sd, present_particle_types[0]).masses = \
        np.ones(n_g, dtype=float) * 1e3 * u.msun
    getattr(sd, present_particle_types[0]).internal_energy = \
        np.ones(n_g, dtype=float) * 1e4 * u.kb * u.K / (1e3 * u.msun)
    getattr(sd, present_particle_types[0]).generate_smoothing_lengths(
        boxsize=boxsize * u.Mpc, dimension=3
    )
    sd.write(snapfile)  # IDs auto-generated

    # Add a named columns dataset and cells info.
    with h5py.File(snapfile, "r+") as f:
        g = f.create_group("Cells")
        g.create_dataset("Centres", data=np.array([[5, 5, 5]], dtype=float))
        cg = g.create_group("Counts")
        cg.create_dataset("PartType0", data=np.array([n_g]), dtype=int)
        fg = g.create_group("Files")
        fg.create_dataset("PartType0", data=np.array([0], dtype=int))
        mdg = g.create_group("Meta-data")
        mdg.attrs["dimension"] = np.array([[1, 1, 1]], dtype=int)
        mdg.attrs["nr_cells"] = np.array([1], dtype=int)
        mdg.attrs["size"] = np.array([boxsize, boxsize, boxsize], dtype=int)
        og = g.create_group("OffsetsInFile")
        og.create_dataset("PartType0", data=np.array([0], dtype=int))
        hsg = f.create_group("HydroScheme")
        hsg.attrs["Adiabatic index"] = 5.0 / 3.0

        ssg = f.create_group("SubgridScheme")
        ncg = ssg.create_group("NamedColumns")
        ncg.create_dataset(
            "HydrogenIonizationFractions",
            data=np.array([b"Neutral", b"Ionized"], dtype="|S32"),
        )
        g = f["PartType0"]
        f_neutral = np.random.rand(n_g)
        f_ion = 1 - f_neutral
        hifd = g.create_dataset(
            "HydrogenIonizationFractions",
            data=np.array([f_neutral, f_ion], dtype=float).T,
        )
        hifd.attrs[
            "Conversion factor to CGS" " (not including cosmological corrections)"
        ] = np.array([1.0], dtype=float)
        hifd.attrs[
            "Conversion factor to physical CGS" " (including cosmological corrections)"
        ] = np.array([1.0], dtype=float)
        hifd.attrs["U_I exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["U_L exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["U_M exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["U_T exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["U_t exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["a-scale exponent"] = np.array([0.0], dtype=float)
        hifd.attrs["h-scale exponent"] = np.array([0.0], dtype=float)

    return


create_toysnap(toysnap_filename)
sd = SWIFTDataset(toysnap_filename)
sm = SWIFTMask(sd.metadata)
sm.constrain_spatial(np.array([[5,6],[5,6],[5,6]]) * u.Mpc)
sd_masked = SWIFTDataset(toysnap_filename, mask=sm)
sd_masked.gas.hydrogen_ionization_fractions.neutral
```

I've attempted various work-around solutions. One results in stochastic segmentation faults instead of consistent segmentation faults, which is arguably worse, but you can see it at: https://github.com/kyleaoman/swiftsimio/tree/segfault_bugfix. The minimal example of the attempted workaround implemented there is:

```python
import h5py
import numpy as np

h5_filename = "bug.hdf5"
N = 16

with h5py.File(h5_filename, "w") as f:
    data = np.random.rand(2 * N).reshape((N, 2))
    f.create_dataset(
        "dset",
        data=data,
        dtype=float
    )

# work around:
with h5py.File(h5_filename, "r") as f:
    handle = f["/dset"]
    output = np.empty((N, 1), dtype=float)
    source_sel = np.s_[0:N, 0:1]
    dest_sel = np.s_[0:N, 0:1]
    print(source_sel, dest_sel)
    assert output.flags["C_CONTIGUOUS"]
    assert handle[source_sel].shape == output[dest_sel].shape
    handle.read_direct(output, source_sel=source_sel, dest_sel=dest_sel)
    output = output.squeeze(axis=1)
```

I don't know what the best approach to resolving this is - require the fixed `h5py` package once it's released, maybe?

**Comment by bwvdnbro** [comment]
I see, thanks for the information! I have been chasing this bug for a while last week, but then gave up once I convinced myself it was `h5py` related.

Can we revert to an older `h5py` module until this is fixed? After that, it would indeed make sense to require a more recent version.

**Comment by JBorrow** [comment]
I'm not sure that older versions of `h5py` include support for the virtual datasets we need for the distributed snapshots.

**Comment by kyleaoman** [comment]
Looks like releases for h5py >3.1.0 are not happening for python3.6, but swiftsimio now doesn't support python3.6, and the issue is solved in h5py >=3.2.0. I guess this issue can be closed? Unless you want to explicitly require h5py>3.2.0, but that's probably overkill?
